### DISCUSSION TITLE
How should I go about debugging my cloud-config?

### DISCUSSION THREAD

**Discussion by sdwilsh** [kairos-io/kairos#708] [Q&A] [answered]
I'm having a hard time getting my cloud-config working correctly.  The [documentation](https://kairos.io/docs/reference/configuration/#test-your-cloud-configs) helped to a degree, but not fully.

# Background
* I'm following the [manual installation steps](https://kairos.io/docs/installation/manual/) with terminal access on the machine (I cannot use the QR code with my setup).
* I'm writing the config file on a different machine, but I get it to the machine I'm doing the manual installation with by spinning up a temporary python web server (`python3 -m http.server 9000`) and then using `curl` to get the file since passing a URL does not seem to work with `kairos-agent manual-install`.
* The ISO I'm using is [`kairos-ubuntu-22-lts-v1.4.0-k3sv1.26.0+k3s1.iso`]([ubuntu-22-lts](https://github.com/kairos-io/provider-kairos/releases/download/v1.4.0/kairos-ubuntu-22-lts-v1.4.0-k3sv1.26.0+k3s1.iso))
* This is the first time I've ever used cloud-init, yip, or elemental, so I have some serious knowledge gaps that could be an issue.

# Issue
I'm trying to setup static networking, as described in [the documentation for systemd-networkd](https://kairos.io/docs/advanced/networking/), but it is not applying, so I assume there's something wrong with my cloud-config.  The [documentation](https://kairos.io/docs/reference/configuration/#test-your-cloud-configs) suggests I grab an image and then run a docker command to test it.  However, there are a few things that are not clear:
1) Where do I get the right image?  From the documentation, I think I'd go to https://quay.io/repository/kairos/core-ubuntu-22-lts-k3s?tab=tags based on the `quay.io/kairos/core-alpine` image being used in the example and the image I'm using, but I'm not certain that is correct, in part because there is no tag that corresponds to `v1.26.0+k3s1` (the newest appears to be `v1.25.2-k3s1`).
2) The documentation for testing cloud-init shows multiple files, that I assume are of the cloud-init syntax that are being tested, however the manual installation documentation says I can only use a single file when invoking `kairos-agent manual-install`.  Is it sufficient to place my single file and run the `docker` command with the right image?

**Comment by paynejacob** [comment, marked as answer]
1. Regarding the image you will want to use the https://quay.io/repository/kairos/kairos-ubuntu-22-lts repo.  The one mentioned in your post is a core image.  These images do not contain k3s, though for testing yip configuration they should both work.  
2. You can have multiple cloud init files for executing yip stages however Kairos will only read configuration from the configuration file with the highest precedence.  This only  matters if you are configuring settings specific to kairos.  All yip stages will be executed in all cloud init files.

**Reply by sdwilsh** [reply]
Related to (1), is the [documentation](https://kairos.io/docs/reference/configuration/#test-your-cloud-configs) incorrect because it uses a core image (for alpine, in this case) as well?  Also, is the ubuntu image misnamed?  I had assumed it contained k3s because it's in the image name (`core-ubuntu-22-lts-k3s`).

**Reply by paynejacob** [reply]
the `core-ubuntu-22-lts-k3s` image is the provider image, that only contains k3s and it designed for usage by 3rd party provisioning tools. It sounds like we need to add some clarification to the docs about all the different image types.

**Reply by sdwilsh** [reply]
There is a note on the [quickstart](https://kairos.io/docs/getting-started/) page, but I hadn't realized how that related to testing the cloud-config.

This is probably because I'm very new to the project, but I image that perspective is useful for y'all to improve the docs for new folks, so I'll keep posting about some of my issues in the hopes that others who are new and finding this project can have fewer questions :D

**Reply by paynejacob** [reply]
In all honesty you could probably test 99% of cloud-config with any image but it is always nice to be consistent fort consistency's sake. 

Please keep them coming, discussions like this are immensely valuable for us as maintainers and like you said you likely won't be the last to ask them.
